File: 033_MaximumDepthOfBinaryTree.py
"""
Topic: Tree / DFS
Maximum Depth of Binary Tree: recursive depth-first search
Time Complexity: O(n)
Space Complexity: O(h) recursion
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def max_depth(root: Optional[TreeNode]) -> int:
    if not root: return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))

if __name__ == '__main__':
    # tree: [3,9,20,null,null,15,7]
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(max_depth(root))  # 3
    print(max_depth(None))  # 0

File: 033_MaximumDepthOfBinaryTree.cpp
// Topic: Binary Tree / DFS
// Maximum Depth of Binary Tree: Find the number of nodes along the longest path from the root to a
// leaf. Time Complexity: O(n) Space Complexity: O(h) recursion

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int maxDepth(TreeNode* root)
{
    if (!root)
        return 0;
    return 1 + max(maxDepth(root->left), maxDepth(root->right));
}

int main()
{
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    cout << maxDepth(root) << "\n";     // 3
    cout << maxDepth(nullptr) << "\n";  // 0
    return 0;
}
